### PR TITLE
Enable tracking through rectangular arrays

### PR DESCRIPTION
**Pull request by elliottbiondo**
This MR modifies OrangeTrackView to use the TrackerVisitor added in #959. This change does not impart any significant overhead, as evidenced by the plot below; running with pseudo array with and without static dispatch has no appreciable effect on timing.

Screen Shot 2023-10-11 at 3 58 38 PM

### CODE REVIEW

File: src/orange/OrangeTrackView.hh [hunk 1]
```diff
@@ -283,7 +283,11 @@ OrangeTrackView::operator=(Initializer_t const& init)
         lsa.dir() = local.dir;
         lsa.universe() = uid;
 
-        daughter_id = tracker.daughter(tinit.volume);
+        // daughter_id = tracker.daughter(tinit.volume);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]:
```suggestion
```

File: src/orange/OrangeTrackView.hh [hunk 2]
```diff
@@ -283,7 +283,11 @@ OrangeTrackView::operator=(Initializer_t const& init)
...
+        CELER_ASSERT(tinit.volume);
+        visit_tracker([&tinit, &daughter_id](
+                          auto&& t) { daughter_id = t.daughter(tinit.volume); },
+                      uid);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: You should be able to write this as 
```suggestion
        daughter_id  = visit_tracker([&tinit](auto&& t) { return t.daughter(tinit.volume); }, uid);
```
which more closely follows the original

File: src/orange/OrangeTrackView.hh [hunk 3]
```diff
@@ -446,8 +450,13 @@ CELER_FUNCTION Propagation OrangeTrackView::find_next_step()
         return {0, true};
     }
 
-    auto tracker = this->make_tracker(UniverseId{0});
-    auto isect = tracker.intersect(this->make_local_state(LevelId{0}));
+    TrackerVisitor visit_tracker{params_};
+    auto local_state = this->make_local_state(LevelId{0});
+
+    auto isect = visit_tracker(
+        [&local_state](auto&& t) { return t.intersect(local_state); },
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Scope the temporary data to the lambda:
```suggestion
    auto isect = visit_tracker(
        [local_state = this->make_local_state(LevelId{0})](auto&& t) { return t.intersect(local_state); },
```

File: src/orange/OrangeTrackView.hh [hunk 4]
```diff
@@ -485,9 +494,14 @@ CELER_FUNCTION Propagation OrangeTrackView::find_next_step(real_type max_step)
 
     if (!this->has_next_step())
     {
-        auto tracker = this->make_tracker(UniverseId{0});
-        auto isect
-            = tracker.intersect(this->make_local_state(LevelId{0}), max_step);
+        TrackerVisitor visit_tracker{params_};
+        auto local_state = this->make_local_state(LevelId{0});
+        auto isect = visit_tracker(
+            [&local_state, &max_step](auto&& t) {
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Same

File: src/orange/OrangeTrackView.hh [hunk 5]
```diff
@@ -568,7 +582,6 @@ CELER_FUNCTION void OrangeTrackView::move_internal(Real3 const& pos)
         auto lsa = this->make_lsa(lev);
         lsa.pos() = local_pos;
 
-        // Apply "transform down" based on stored transform
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: This comment should still be valid; I guess this got deleted in a rebase?

File: src/orange/OrangeTrackView.hh [hunk 6]
```diff
@@ -651,7 +665,13 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
     size_type level = sl.get();
     LocalVolumeId volume_id = tinit.volume;
     auto universe_id = lsa.universe();
-    auto daughter_id = tracker.daughter(volume_id);
+
+    // auto daughter_id = tracker.daughter(volume_id);
+    DaughterId daughter_id;
+    CELER_ASSERT(volume_id);
+    visit_tracker([&volume_id, &daughter_id](
+                      auto&& t) { daughter_id = t.daughter(volume_id); },
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Same as earlier comment; delete the dead code and return by value

File: src/orange/OrangeTrackView.hh [hunk 7]
```diff
@@ -668,10 +688,14 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
         // XXX I think the next line is redundant
         local.temp_sense = this->make_temp_sense();
 
-        // Initialize in daughter and get IDs of volume and potential daughter
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh:671]: Restore comment

File: src/orange/OrangeTrackView.hh [hunk 8]
```diff
@@ -706,10 +730,16 @@ CELER_FUNCTION void OrangeTrackView::set_dir(Real3 const& newdir)
...
         // Changing direction on a boundary is dangerous, as it could mean we
         // don't leave the volume after all.
-        auto tracker = this->make_tracker(UniverseId{0});
         detail::UniverseIndexer ui(params_.universe_indexer_data);
-        Real3 normal = tracker.normal(
-            this->pos(), ui.local_surface(this->surface_id()).surface);
+
+        TrackerVisitor visit_tracker{params_};
+        auto pos = this->pos();
+        auto local_surface = ui.local_surface(this->surface_id()).surface;
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Make the temporary data local

File: src/orange/OrangeTrackView.hh [hunk 9]
```diff
@@ -880,9 +910,15 @@ OrangeTrackView::find_next_step_impl(detail::Intersection isect)
...
         auto lsa = this->make_lsa(levelid);
-        auto tracker = this->make_tracker(lsa.universe());
-        auto local_isect = tracker.intersect(this->make_local_state(levelid),
-                                             isect.distance);
+
+        TrackerVisitor visit_tracker{params_};
+        auto local_state = this->make_local_state(levelid);
+        auto local_isect = visit_tracker(
+            [&local_state, &isect](auto&& t) {
+                return t.intersect(local_state, isect.distance);
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh]: Same as other comments

File: src/orange/OrangeTrackView.hh [hunk 10]
```diff
@@ -918,8 +954,11 @@ CELER_FUNCTION real_type OrangeTrackView::find_safety()
     auto lsa = this->make_lsa();
 
     CELER_ASSERT(lsa.universe() == UniverseId{0});
-    auto tracker = this->make_tracker(lsa.universe());
-    return tracker.safety(lsa.pos(), lsa.vol());
+
+    TrackerVisitor visit_tracker{params_};
+    return visit_tracker(
```
Comment (sethrj) [inline src/orange/OrangeTrackView.hh:952]: I wonder if it's worth having a `template<class F> auto visit_tracker(F&&, UniverseId) const;` (and same for `visit_transform`)?

**Review by sethrj — CHANGES_REQUESTED** [review]

File: src/orange/OrangeTrackView.hh (hunk 10 above)
Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh:952]: You mean something like:

```
template<class F> decltype(auto) OrangeTrackView::visit_tracker(F&& f, UniverseId uid)
{
        TrackerVisitor visit{params_};
        return visit(f, uid);
}
```
?

Comment (elliottbiondo) [reply src/orange/OrangeTrackView.hh:952]: @sethrj, I tried the code above, but I am getting type deduction errors, despite this the same API as what is done within TrackerVisit. The rest of  this PR is good to go.

Comment (sethrj) [reply src/orange/OrangeTrackView.hh:952]: Thanks for trying that out. We can simplify the code in the future if needed.

**Review by sethrj — APPROVED** [review]
Nice work!

**Pull request merged**